compiler/util/kernel_type.py

```diff
@@ -0,0 +1,240 @@
+        """
+        Parses a zero point adjustment operation and returns the two operands.
+        A zero point adjustment operation is a subtraction operation, where
+        the right operand is a zero point offset and the left operand is the
+        input tensor. Before the subtraction, the input tensor may be sign
+        extended to avoid overflow.
+
+        Args:
+            op (Operation): The subtraction operation.
+
+        Returns:
+            tuple[Operation | SSAValue, Operation | SSAValue]: The two operands
+            of the subtraction operation.
+
+        Raises:
+            KernelException: If the provided operation is not a subtraction.
+        """
+
+        # check if op is a subtraction
+        if not isinstance(op, arith.Subi):
+            raise KernelException
+
+        # get operands
+        value = op.lhs
+        adjustment = op.rhs
+
+        # additional sign extension possible
+        while isinstance(value, OpResult) and isinstance(value.op, arith.ExtSIOp):
+            value = value.op.operands[0]
+
+        return value, adjustment
+
+    @staticmethod
+    def parse_inputs(linalg_op: linalg.Generic) -> tuple[linalg.YieldOp, dict]:
+        """
+        Parses the inputs of a linalg operation and returns the yield operation
+        of the linalg body and a dictionary of input types. The dictionary of
+        input types is used to match the block arguments (scalars) to the actual
+        input types of the linalg.Generic (shaped memrefs).
+
+        Args:
+            linalg_op (linalg.Generic): The linalg operation.
+
+        Returns:
+            tuple[linalg.YieldOp, dict]: The yield operation and a
+            dictionary of input types.
+
+        Raises:
+            KernelException: If the number of shaped inputs is incorrect for binary ops.
+            KernelException: If the last op of the linalg block is not a yield op.
+            KernelException: If the yield op does not return exactly one result.
+            KernelException: If the yield op operand is not the result of another op.
+        """
+
+        linalg_block: Block = linalg_op.body.block
+        input_types: list = [o.type for o in linalg_op.operands]
+
+        # the inputs of the linalg block are scalars, but to detect
+        # the kernel type, we need to know if the inputs of the linalg
+        # block are tensors or scalars, therefore we generate a dictionary
+        # of the input types of the linalg
+        types = dict(zip(linalg_block.args, input_types))
+
+        # this check counts the number of inputs which are shaped inputs
+        # for all the kernels there may be three shaped operands
+        # (input a, input b, output c)  other inputs may be integer types used for
+        # constant values, such as zero-point offsets
+        if len([type for type in types.values() if isinstance(type, ShapedType)]) != 3:
+            raise KernelException("Wrong number of shaped inputs")
```

Comment:
Maybe make this more explicit?
```suggestion
            raise KernelException("Wrong number of shaped inputs: Expect 2 shaped inputs and 1 shaped output")
```